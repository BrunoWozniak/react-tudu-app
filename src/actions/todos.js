import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_TODO
export const addTodo = (todo) => ({
    type: 'ADD_TODO',
    todo
});

export const startAddTodo = (todoData = {}) => {
    return (dispatch, getState) => {
        const {
            note = '',
            description = '',
            amount = 0,
            createdAt = 0
        } = todoData;
        const todo = {description, note, amount, createdAt};
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/todos`).push(todo).then((ref) => {
            dispatch(addTodo({
                id: ref.key,
                ...todo
            }));
        });
    };
};

// REMOVE_TODO
export const removeTodo = ({ id } = {}) => ({
    type: 'REMOVE_TODO',
    id
});

export const startRemoveTodo = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/todos/${id}`).remove().then(() => {
            dispatch(removeTodo({ id }));
        });
    };
};

// EDIT_TODO
export const editTodo = (id, updates) => ({
    type: 'EDIT_TODO',
    id,
    updates
});

export const startEditTodo = (id, updates) => {
    return (dispatch, getState) => {       
        const uid = getState().auth.uid; 
        return database.ref(`users/${uid}/todos/${id}`).update(updates).then(() => {
            dispatch(editTodo(id, updates));
        });
    };
};

// SET_TODOS
export const setTodos = (todos) => ({
    type: 'SET_TODOS',
    todos
});

export const startSetTodos = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/todos`).once('value').then((snapshot) => {
            const fetchedTodos = [];
            snapshot.forEach((snapshotItem) => {
                fetchedTodos.push({
                    id: snapshotItem.key,
                    ...snapshotItem.val()
                });
            });
            dispatch(setTodos(fetchedTodos));
        }).catch((e) => {
            console.log('Error!', e);
        });
    };
};