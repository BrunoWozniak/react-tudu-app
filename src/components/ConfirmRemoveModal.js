import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';

export default class ConfirmRemoveModal extends React.Component {
    constructor(props) {
        super(props);
    }
    
    render () {
        return(
            <Modal
                isOpen={!!this.props.confirmDelete}
                contentLabel="Are you sure?"
                closeTimeoutMS={200}
                className="modal"
            >
                <h3 className="modal__title">Are you sure?</h3>
                <p className="modal__body">Please confirm you want to delete this expense.</p>
                <div className="modal__buttons">
                    <button className="button" id="btnRemoveConfirm" onClick={this.props.onConfirmDelete}> Confirm </button>
                    <button className="button button--secondary" id="btnRemoveCancel" onClick={this.props.onCancelDelete} > Cancel </button>
                </div>
            </Modal>
        );
    }
}