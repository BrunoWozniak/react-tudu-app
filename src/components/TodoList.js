import React from 'react';
import { connect } from 'react-redux';
import TodoListItem from './TodoListItem';
import selectTodos from '../selectors/todos';

export const TodoList = (props) => (
	<div className="content-container">
		<div className="list-header">
			<div className="show-on-mobile">Todos</div>
			<div className="show-on-desktop">Todo</div>
			<div className="show-on-desktop">Amount</div>
		</div>
		<div className="list-body">
			{
				props.todos.length === 0 ? (
					<div className="list-item list-item--message">
						<span>No todo</span>
					</div>
				) : (
					props.todos.map((todo) => {
						return <TodoListItem key={todo.id} {...todo} />
					})
				)
			}
		</div>
	</div>
);

const mapStateToProps = (state) => {
	return {
		todos: selectTodos(state.todos, state.filters)
	};
};

export default connect(mapStateToProps)(TodoList);