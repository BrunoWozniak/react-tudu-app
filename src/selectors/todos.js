import moment from 'moment';

// Get visible todos

export default (todos, { searchText, sortBy, startDate, endDate }) => {
	return todos.filter(({ dueAt, text }) => {
		const dueAtMoment = moment(dueAt);
		const startDateMatch = startDate ? startDate.isSameOrBefore(dueAtMoment, 'day') : true;
		const endDateMatch = endDate ? endDate.isSameOrAfter(dueAtMoment, 'day') : true;
		const textMatch = text.toLowerCase().includes(searchText.toLowerCase());

		return startDateMatch && endDateMatch && textMatch;
	}).sort((a, b) => {
		if (sortBy === 'descending') {
			return a.dueAt < b.dueAt ? 1 : -1;
		} else if (sortBy === 'ascending') {
			return a.dueAt > b.dueAt ? 1 : -1;
		}
	});
};