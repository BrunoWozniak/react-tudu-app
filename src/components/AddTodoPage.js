import React from 'react';
import { connect } from 'react-redux';
import TodoForm from './TodoForm';
import { startAddTodo } from '../actions/todos';


export class AddTodoPage extends React.Component {
    onSubmit = (todo) => {
        this.props.startAddTodo(todo);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add a Todo</h1>

                    </div>
                </div>
                <div className="content-container">
                    <TodoForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddTodo: (todo) => dispatch(startAddTodo(todo))
});

export default connect(undefined, mapDispatchToProps)(AddTodoPage);