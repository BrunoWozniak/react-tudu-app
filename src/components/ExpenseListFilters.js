import React from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDateFilter, setEndDateFilter } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }
  
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  onShowAll = () => {
    this.props.setStartDate(null);
    this.props.setEndDate(null);
    this.props.setTextFilter('');
  }
  
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
          <div>
            <button
              className="button button--filter-bar"
              onClick={this.onShowAll}
            >
              Show all
            </button>
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDateFilter(startDate)),
  setEndDate: (endDate) => dispatch(setEndDateFilter(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
