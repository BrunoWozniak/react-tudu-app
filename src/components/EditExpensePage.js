import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ConfirmRemoveModal from './ConfirmRemoveModal';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            confirmDelete: props.confirmDelete
        };
    };

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.setState({ confirmDelete: true });
    }

    onConfirmDelete = (expense) => {
        this.setState({ confirmDelete: false });
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    onCancelDelete = () => {
        this.setState({ confirmDelete: false })
    }
    
    render() {
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>
                        Remove Expense
                    </button>
                </div>
                <ConfirmRemoveModal
                    confirmDelete={this.state.confirmDelete}
                    onCancelDelete={this.onCancelDelete}
                    onConfirmDelete={this.onConfirmDelete}
                />
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
    confirmDelete: undefined
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);