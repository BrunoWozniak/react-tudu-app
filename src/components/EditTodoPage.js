import React from 'react';
import { connect } from 'react-redux';
import TodoForm from './TodoForm';
import { startEditTodo, startRemoveTodo } from '../actions/todos';
import ConfirmRemoveModal from './ConfirmRemoveModal';

export class EditTodoPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            confirmDelete: props.confirmDelete
        };
    };

    onSubmit = (todo) => {
        this.props.startEditTodo(this.props.todo.id, todo);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.setState({ confirmDelete: true });
    }

    onConfirmDelete = (todo) => {
        this.setState({ confirmDelete: false });
        this.props.startRemoveTodo({ id: this.props.todo.id });
        this.props.history.push('/');
    }

    onCancelDelete = () => {
        this.setState({ confirmDelete: false })
    }
    
    render() {
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Todo</h1>
                    </div>
                </div>
                <div className="content-container">
                    <TodoForm
                        todo={this.props.todo}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>
                        Remove Todo
                    </button>
                </div>
                <ConfirmRemoveModal
                    confirmDelete={this.state.confirmDelete}
                    onCancelDelete={this.onCancelDelete}
                    onConfirmDelete={this.onConfirmDelete}
                />
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    todo: state.todos.find((todo) => todo.id === props.match.params.id),
    confirmDelete: undefined
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditTodo: (id, todo) => dispatch(startEditTodo(id, todo)),
    startRemoveTodo: (id) => dispatch(startRemoveTodo(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditTodoPage);