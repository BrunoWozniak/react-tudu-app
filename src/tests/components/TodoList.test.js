import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../components/TodoList';
import todos from '../fixtures/todos';

test('should render TodoList with todos', () => {
    const wrapper = shallow(<TodoList todos={todos}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render TodoList with empty todo list', () => {
    const wrapper = shallow(<TodoList todos={[]}/>);
    expect(wrapper).toMatchSnapshot();
});