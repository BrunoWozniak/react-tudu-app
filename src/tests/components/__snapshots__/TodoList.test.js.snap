// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TodoList with empty todo list 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-on-mobile"
    >
      Todos
    </div>
    <div
      className="show-on-desktop"
    >
      Todo
    </div>
    <div
      className="show-on-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No todo
      </span>
    </div>
  </div>
</div>
`;

exports[`should render TodoList with todos 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-on-mobile"
    >
      Todos
    </div>
    <div
      className="show-on-desktop"
    >
      Todo
    </div>
    <div
      className="show-on-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <TodoListItem
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <TodoListItem
      amount={109500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <TodoListItem
      amount={4500}
      createdAt={345600000}
      description="Credit card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
